app/models/action/edition_diff_factory.rb

```diff
@@ -0,0 +1,48 @@
+class Action::EditionDiffFactory
+  def add_previous_edition(previous_edition)
+    @previous_edition = get_edition_attributes(previous_edition)
+  end
+
+  def add_updated_edition(updated_edition)
+    @updated_edition = get_edition_attributes(updated_edition)
+  end
+
+  def diff
+    raise MissingEditionError, 'No previous_edition provided' if @previous_edition.nil?
+    raise MissingEditionError, 'No updated_edition provided' unless @updated_edition.present?
+    create_edition_diff(@previous_edition, @updated_edition)
+  end
+
+private
+
+  def create_edition_diff(edition_a, edition_b)
+    HashDiff.diff(edition_a, edition_b)
+  end
+
+  def get_edition_attributes(edition)
+    attributes = {}
+    return attributes unless edition.present?
+
+    Edition::TOP_LEVEL_FIELDS.each do |field|
+      attributes[field.to_s] = edition.send(field)
+    end
+
+    if edition.links.present?
+      attributes["links"] = {}
+      edition.links.each do |link|
+        attributes["links"][link["link_type"]] = []
+      end
+      edition.links.each do |link|
+        attributes["links"][link["link_type"]] << link["target_content_id"]
+      end
```

Comment:
I'd probably go with an `each_with_object` here, you can then have that result of that be set to attributes["links"] and can avoid iterating the collection twice. 

Something like
```ruby
attributes["links"] = edition.links.each_with_object({}) do |link, memo|
  memo[link.link_type] ||= []
  memo[link.link_type] << link.target_content_id
end
```

You also don't need to check if links are present, it's better we have them as an empty hash than missing from the collection